keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md

```diff
@@ -0,0 +1,293 @@
+* Support kubeadm join for Windows
+* Support kubeadm reset for Windows
+* Create and maintain a script to deploy WIndows CNIs
+
+### Non-Goals
+
+* Installing the Container Runtime (e.g. Docker or containerd)
+* Implement kubeadm init for Windows
+* Implement kubeadm join --control-plane
+* Supporting upgrades using kubeadm upgrade
+* Running kube-proxy as a DaemonSet on Windows
+* Running Flannel as a DaemonSet on Windows
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+A user will have a Windows machine that they want to join to an existing Kubernetes cluster.
+
+To start, the user will download a set of required binaries (kubelet, kube-proxy, kubeadm, kubectl, Flannel) using a script. The same script will also wrap kubeadm execution.
+
+After executing the script, the kubelet and kube-proxy will be registered as Windows services.
+
+Next, the script will run "kubeadm join ..." to join the node to the cluster. In this step kubeadm will run preflight checks and proceed with the regular join process.
+
+kubeadm will then restart the kubelet service using flags that kubeadm fed to the Windows service. kubeadm will proceed to bootstrap the node. After this process is finished the node should show with the status of NotReady.
+
+The same script will then configure FlannelD. The script will (re-)register FlannelD as a service and start it with the correct configuration, optionally using parameters from kubeadm.
+
+kube-proxy will do the same steps as FlannelD shortly after.
+
+After these steps, the node status should be Ready.
+
+### Implementation Details/Notes/Constraints
+
+#### Kubeadm manages the kubelet start / stop as a service
+
+Kubeadm has a “kubelet” phase that can start/stop the [kubelet service](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/kubelet.go).
+
+Kubeadm already supports running the kubelet as a Windows service. This support was added by these two PRs: [60144](https://github.com/kubernetes/kubernetes/pull/60144), [60144](https://github.com/kubernetes/kubernetes/pull/53553).
+
+This existing support will be leveraged by this proposal.
+
+#### Kubeadm makes assumptions about systemd and Linux
+
+Kubeadm is quite coupled to systemd and Linux at the moment.
+This file shows some of the kubelet-flags that are Linux bound:
+[kubelet flags](https://github.com/kubernetes/kubernetes/blob/b8b689aae0d6879de5192b203590330a11c7b9e3/cmd/kubeadm/app/phases/kubelet/flags.go)
+
+The following changes are required::
+* Implement proper cri-socket detection in kubeadm for Windows.
+* Omit flags such as “cgroup-driver”
+* Specify the correct “pod-infra-container-image”
+* Also omit writing systemd drop-in files on Windows in general.
+* Populate the Windows service dynamically.
+
+To support Windows specific flags for the kubelet, there is a requirement to split kubeadm’s app/phases/kubelet/flags.go files into two:
+* app/phases/kubelet/flags_windows.go
+* app/phases/kubelet/flags_linux.go
```

Comment:
I'd like work from the componentconfig folks.

Reply:
could you elaborate on your ideas?
this could end up being something quite isolated to kubeadm -> kubelet. 
`if os == "windows" { use flagset A } else { use flag set B }`

Reply:
You could build meta-data in apis ~= +omitempty

Reply:
At the current state of kubelet's component config (which is at v1beta1), there are a few command line flags that don't have a corresponding field in the config. Those are:

- dockershim related flags
- `container-runtime` && `container-runtime-endpoint` flags
- `register-with-taints` flag
- `hostname-override`

Only the `resolv-conf` flag has a representation in the component config. However, we would have to _patch_ it, after fetching the config from the config map, for the local machine setting to take place.